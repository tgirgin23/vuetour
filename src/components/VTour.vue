<template>
  <div class="v-tour">
    <v-mask
      v-if="backgroundMask"
      :windowHeight="this.mask.windowHeight"
      :windowWidth="this.mask.windowWidth"
      :targetWidth="this.mask.targetWidth"
      :targetHeight="this.mask.targetHeight"
      :targetLeft="this.mask.targetLeft"
      :targetTop="this.mask.targetTop"
    ></v-mask>
    <v-step
      :id="'v-step-' + stepHash"
      :ref="'v-step-' + stepHash"
      v-if="steps[currentStep]"
      :stepHash="stepHash"
      :stepParams="stepParams"
      :step="steps[currentStep]"
      :key="currentStep"
      :previous-step="previousStep"
      :next-step="nextStep"
      :stop="stop"
      :skip="skip"
      :finish="finish"
      :is-first="isFirst"
      :is-last="isLast"
      :labels="customOptions.labels"
      :enabled-buttons="customOptions.enabledButtons"
      :highlight="customOptions.highlight"
      :stop-on-fail="customOptions.stopOnTargetNotFound"
      :debug="customOptions.debug"
      :targetElement="targetElement"
      @targetNotFound="$emit('targetNotFound', $event)"
      @show-mask="showMask"
    >
    </v-step>
  </div>
</template>

<script>
import { DEFAULT_CALLBACKS, DEFAULT_OPTIONS, DEFAULT_STEP_OPTIONS, KEYS } from '@/shared/constants'
import VMask from '@/components/VMask'
import { customAlphabet } from 'nanoid'

export default {
  name: 'v-tour',
  components: { VMask },
  props: {
    steps: {
      type: Array,
      default: () => []
    },
    name: {
      type: String
    },
    options: {
      type: Object,
      default: () => { return DEFAULT_OPTIONS }
    },
    callbacks: {
      type: Object,
      default: () => { return DEFAULT_CALLBACKS }
    }
  },
  data () {
    return {
      currentStep: -1,
      backgroundMask: false,
      mask: {
        windowHeight: 0,
        windowWidth: 0,
        targetHeight: 0,
        targetWidth: 0,
        targetTop: 0,
        targetLeft: 0
      },
      targetElement: null,
      stepHash: null
    }
  },
  mounted () {
    this.$tours[this.name] = this
    if (this.customOptions.useKeyboardNavigation) {
      window.addEventListener('keyup', this.handleKeyup)
    }

    this.mask.windowWidth = window.innerWidth
    this.mask.windowHeight = document.body.scrollHeight

    window.addEventListener('resize', this.handleWindowResize)
  },
  beforeDestroy () {
    // Remove the keyup listener if it has been defined
    if (this.customOptions.useKeyboardNavigation) {
      window.removeEventListener('keyup', this.handleKeyup)
    }

    window.removeEventListener('resize', this.handleWindowResize)
  },
  computed: {
    // Allow us to define custom options and merge them with the default options.
    // Since options is a computed property, it is reactive and can be updated during runtime.
    customOptions () {
      return {
        ...DEFAULT_OPTIONS,
        ...this.options
      }
    },
    stepParams () {
      return {
        ...DEFAULT_STEP_OPTIONS,
        ...{ highlight: this.highlight }, // Use global tour highlight setting first
        ...{ enabledButtons: Object.assign({}, this.enabledButtons) },
        ...this.step.params // Then use local step parameters if defined
      }
    },
    customCallbacks () {
      return {
        ...DEFAULT_CALLBACKS,
        ...this.callbacks
      }
    },
    // Return true if the tour is active, which means that there's a VStep displayed
    isRunning () {
      return this.currentStep > -1 && this.currentStep < this.numberOfSteps
    },
    isFirst () {
      return this.currentStep === 0
    },
    isLast () {
      return this.currentStep === this.steps.length - 1
    },
    numberOfSteps () {
      return this.steps.length
    },
    step () {
      return this.steps[this.currentStep]
    }
  },
  methods: {
    async start (startStep) {
      // Wait for the DOM to be loaded, then start the tour
      startStep = typeof startStep !== 'undefined' ? parseInt(startStep, 10) : 0
      let step = this.steps[startStep]

      let process = () => new Promise((resolve) => {
        setTimeout(() => {
          this.customCallbacks.onStart()
          this.currentStep = startStep
          resolve()
        }, this.customOptions.startTimeout)
      })

      if (typeof step.before !== 'undefined') {
        try {
          await step.before('start')
        } catch (e) {
          return Promise.reject(e)
        }
      }
      await process()

      return Promise.resolve()
    },
    async previousStep () {
      let futureStep = this.currentStep - 1

      let process = () => new Promise((resolve) => {
        this.customCallbacks.onPreviousStep(this.currentStep)
        this.currentStep = futureStep
        resolve()
      })

      if (futureStep > -1) {
        let step = this.steps[futureStep]
        if (typeof step.before !== 'undefined') {
          try {
            await step.before('previous')
          } catch (e) {
            return Promise.reject(e)
          }
        }
        await process()
      }

      return Promise.resolve()
    },
    async nextStep () {
      let futureStep = this.currentStep + 1

      let process = () => new Promise((resolve) => {
        this.customCallbacks.onNextStep(this.currentStep)
        this.currentStep = futureStep
        resolve()
      })

      if (futureStep < this.numberOfSteps && this.currentStep !== -1) {
        let step = this.steps[futureStep]
        if (typeof step.before !== 'undefined') {
          try {
            await step.before('next')
          } catch (e) {
            return Promise.reject(e)
          }
        }
        await process()
      }

      return Promise.resolve()
    },
    stop () {
      this.showMask(false)
      this.customCallbacks.onStop()
      document.body.classList.remove('v-tour--active')
      this.currentStep = -1
    },
    skip () {
      this.customCallbacks.onSkip()
      this.stop()
    },
    finish () {
      this.customCallbacks.onFinish()
      this.stop()
    },
    handleKeyup (e) {
      if (this.customOptions.debug) {
        console.log('[Vue Tour] A keyup event occurred: ', e)
      }
      switch (e.keyCode) {
        case KEYS.ARROW_RIGHT:
          this.isKeyEnabled('ARROW_RIGHT') && this.nextStep()
          break
        case KEYS.ARROW_LEFT:
          this.isKeyEnabled('ARROW_LEFT') && this.previousStep()
          break
        case KEYS.ESCAPE:
          this.isKeyEnabled('ESCAPE') && this.stop()
          break
      }
    },
    isKeyEnabled (key) {
      const { enabledNavigationKeys } = this.customOptions
      return enabledNavigationKeys.hasOwnProperty(key) ? enabledNavigationKeys[key] : true
    },
    handleWindowResize () {
      this.mask.windowWidth = window.innerWidth
    },
    showMask (show) {
      document.documentElement.style.overflow = show ? 'hidden' : 'none'
      this.backgroundMask = show
    }
  },
  watch: {
    step () {
      this.stepHash = customAlphabet('1234567890abcdef', 8)()
      this.targetElement = document.querySelector(this.step?.target)
      // If we have a target element we create a mask around it
      if (this.targetElement) {
        const { left, top, height, width } = this.targetElement?.getBoundingClientRect()
        this.mask = Object.assign({}, this.mask, {
          targetHeight: height,
          targetWidth: width,
          targetLeft: left,
          targetTop: top
        })
        this.showMask(this.stepParams.mask)
      } else if (this.step?.params?.mask) {
        this.mask = Object.assign({}, this.mask, {
          targetHeight: 0,
          targetWidth: 0,
          targetLeft: 0,
          targetTop: 0
        })
        this.showMask(this.stepParams.mask)
      }
    }
  }
}
</script>

<style lang="scss">
  body.v-tour--active {
    pointer-events: none;
  }

  .v-tour {
    pointer-events: auto;

    &__target {
      &--highlighted {
        box-shadow: 0 0 0 4px rgba(0,0,0,.4);
        pointer-events: auto;
        z-index: 9999;
      }
      &--relative {
        position: relative;
      }
    }
  }
</style>
