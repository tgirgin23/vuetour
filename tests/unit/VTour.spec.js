import { mount } from '@vue/test-utils'
import Vue from 'vue'
import VueTour from '@/main'
import VTour from '@/components/VTour.vue'
import { JSDOM } from 'jsdom'
import nanoid, { customAlphabet } from 'nanoid'

const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

Vue.use(VueTour)

const steps = [
  {
    target: '#v-step-0',
    content: `Discover <strong>Vue Tour</strong>!`
  },
  {
    target: '#v-step-1',
    content: 'An awesome plugin made with Vue.js!'
  }
]

beforeAll(() => {
  // Mock the scrollIntoView function
  Element.prototype.scrollIntoView = jest.fn()

  // Add div elements mocking the steps
  const docFrag = document.createDocumentFragment()

  steps.forEach(step => {
    const div = document.createElement('div')
    div.setAttribute('id', step.target)
    docFrag.appendChild(div)
  })

  document.body.appendChild(docFrag)
})

describe('VTour.vue', () => {
  it('has the correct number of steps', () => {
    const wrapper = mount(VTour, {
      propsData: {
        name: 'myTestTour',
        steps
      }
    })

    expect(wrapper.vm.steps.length).toEqual(2)

    expect(wrapper.element).toMatchSnapshot()
  })

  it('registers itself in the global Vue instance', () => {
    const wrapper = mount(VTour, {
      propsData: {
        name: 'myTestTour',
        steps: []
      }
    })

    expect(typeof wrapper.vm.$tours).toBe('object')
    expect(wrapper.vm.$tours).toHaveProperty('myTestTour')

    expect(wrapper).toMatchSnapshot()
  })

  it('stays within the boundaries of the number of steps', async () => {
    const wrapper = mount(VTour, {
      propsData: {
        name: 'myTestTour',
        steps
      }
    })

    expect(wrapper.vm.currentStep).toEqual(-1)

    expect(wrapper).toMatchSnapshot()

    await wrapper.vm.start()
    expect(wrapper.vm.currentStep).toEqual(0)

    expect(wrapper).toMatchSnapshot()

    // We call nextStep one more time than needed
    for (let i = 0; i < steps.length; i++) {
      await wrapper.vm.nextStep()
    }

    expect(wrapper.vm.currentStep).toEqual(1)

    expect(wrapper).toMatchSnapshot()

    // We call previousStep one more time than needed
    for (let i = 0; i < steps.length; i++) {
      await wrapper.vm.previousStep()
    }

    expect(wrapper.vm.currentStep).toEqual(0)

    wrapper.vm.stop()

    expect(wrapper.vm.currentStep).toEqual(-1)

    expect(wrapper).toMatchSnapshot()
  })

  describe('#before', () => {
    let step0 = false
    let step1 = false
    const beforeSteps = [
      {
        target: '#v-step-0',
        content: `Discover <strong>Vue Tour</strong>!`,
        before: () => {
          step0 = true
          return Promise.resolve()
        }
      },
      {
        target: '#v-step-1',
        content: 'An awesome plugin made with Vue.js!',
        before: () => {
          step1 = true
          return Promise.resolve()
        }
      },
      {
        target: '#v-step-2',
        content: 'An awesome plugin made with Vue.js!',
        before: () => {
          return Promise.reject(new Error('testing'))
        }
      },
      {
        target: '#v-step-3',
        content: 'An awesome plugin made with Vue.js!',
        before: () => {
          return Promise.resolve()
        }
      }
    ]

    it('invokes before() on start()', async () => {
      const wrapper = mount(VTour, {
        propsData: {
          name: 'myTestTour',
          steps: beforeSteps
        }
      })

      await wrapper.vm.start()
      expect(wrapper.vm.currentStep).toEqual(0)
      expect(step0).toEqual(true)

      jest.mock('nanoid', () => {
        return {
          customAlphabet: jest.fn(() => 1)()
        }
      })

      // nanoid.customAlphabet('1234567890abcdef').mockImplementation(() => 1)

      // jest.mock('nanoid', () => ({ customAlphabet: () => jest.fn()() }))
      // test2 = jest.fn(() => 1)()

      jest.mock(
        'nanoid',
        () => {
          const _customAlphabet = jest.fn().mockReturnValue('1234abcd')
          return {
            customAlphabet: jest.fn(() => _customAlphabet)
          }
        },
        { virtual: true }
      )

      expect(wrapper.element).toMatchSnapshot()

      step0 = false
      step1 = false
    })

    it('invokes before() on nextStep()', async () => {
      const wrapper = mount(VTour, {
        propsData: {
          name: 'myTestTour',
          steps: beforeSteps
        }
      })

      await wrapper.vm.start()
      expect(wrapper.vm.currentStep).toEqual(0)

      await wrapper.vm.nextStep()
      expect(wrapper.vm.currentStep).toEqual(1)
      expect(step1).toEqual(true)

      step0 = false
      step1 = false
    })

    it('handles before() promise rejection on start', async () => {
      const wrapper = mount(VTour, {
        propsData: {
          name: 'myTestTour',
          steps: beforeSteps
        }
      })

      try {
        await wrapper.vm.start(2)
        expect(true).toEqual(false) // dead code
      } catch (e) {
        expect(e.message).toEqual('testing')
        expect(wrapper.vm.currentStep).toEqual(-1)
      }
    })

    it('handles before() promise rejection on nextStep()', async () => {
      const wrapper = mount(VTour, {
        propsData: {
          name: 'myTestTour',
          steps: beforeSteps
        }
      })

      await wrapper.vm.start(1)

      try {
        await wrapper.vm.nextStep()
        expect(true).toEqual(false) // dead code
      } catch (e) {
        expect(e.message).toEqual('testing')
        expect(wrapper.vm.currentStep).toEqual(1)
      }
    })

    it('handles before() promise rejection on previousStep()', async () => {
      const wrapper = mount(VTour, {
        propsData: {
          name: 'myTestTour',
          steps: beforeSteps
        }
      })

      await wrapper.vm.start(3)

      try {
        await wrapper.vm.previousStep()
        expect(true).toEqual(false) // dead code
      } catch (e) {
        expect(e.message).toEqual('testing')
        expect(wrapper.vm.currentStep).toEqual(3)
      }
    })
  })
})
