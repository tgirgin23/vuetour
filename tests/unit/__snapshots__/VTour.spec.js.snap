// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VTour.vue #before invokes before() on start() 1`] = `
<div
  class="v-tour"
>
  <!---->
   
  <div
    class="v-step v-step__center"
    id="v-step-37b59aad"
  >
    <!---->
     
    <div
      class="v-step__content"
    >
      <div>
        Discover 
        <strong>
          Vue Tour
        </strong>
        !
      </div>
    </div>
     
    <div
      class="v-step__buttons"
    >
      <button
        class="v-step__button v-step__button-skip"
      >
        Skip tour
      </button>
       
      <!---->
       
      <button
        class="v-step__button v-step__button-next"
      >
        Next
      </button>
       
      <!---->
    </div>
     
    <div
      class="v-step__arrow"
    />
  </div>
</div>
`;

exports[`VTour.vue has the correct number of steps 1`] = `
<div
  class="v-tour"
>
  <!---->
   
  <!---->
</div>
`;

exports[`VTour.vue registers itself in the global Vue instance 1`] = `
<div class="v-tour">
  <!---->
  <!---->
</div>
`;

exports[`VTour.vue stays within the boundaries of the number of steps 1`] = `
<div class="v-tour">
  <!---->
  <!---->
</div>
`;

exports[`VTour.vue stays within the boundaries of the number of steps 2`] = `
<div class="v-tour">
  <!---->
  <div class="v-step v-step__center" id="v-step-4dd0ff2a">
    <!---->
    <div class="v-step__content">
      <div>Discover <strong>Vue Tour</strong>!</div>
    </div>
    <div class="v-step__buttons"><button class="v-step__button v-step__button-skip">Skip tour</button>
      <!----> <button class="v-step__button v-step__button-next">Next</button>
      <!---->
    </div>
    <div class="v-step__arrow"></div>
  </div>
</div>
`;

exports[`VTour.vue stays within the boundaries of the number of steps 3`] = `
<div class="v-tour">
  <!---->
  <div class="v-step v-step__center" id="v-step-9095e0b4">
    <!---->
    <div class="v-step__content">
      <div>An awesome plugin made with Vue.js!</div>
    </div>
    <div class="v-step__buttons">
      <!----> <button class="v-step__button v-step__button-previous">Previous</button>
      <!----> <button class="v-step__button v-step__button-stop">Finish</button>
    </div>
    <div class="v-step__arrow"></div>
  </div>
</div>
`;

exports[`VTour.vue stays within the boundaries of the number of steps 4`] = `
<div class="v-tour">
  <!---->
  <div class="v-step v-step__center" id="v-step-45416562">
    <!---->
    <div class="v-step__content">
      <div>Discover <strong>Vue Tour</strong>!</div>
    </div>
    <div class="v-step__buttons"><button class="v-step__button v-step__button-skip">Skip tour</button>
      <!----> <button class="v-step__button v-step__button-next">Next</button>
      <!---->
    </div>
    <div class="v-step__arrow"></div>
  </div>
</div>
`;
